### ISSUE TITLE
Alpha Zero / MuZero differences

### ISSUE THREAD

**Issue by Counterfeiter** [werner-duvaud/muzero-general#143]
Hello,

I wonder about the slow learning and slow game stepping of my custom environment. I checked the tree search and can't find any game step execution. The value, reward and policy heads could only be trained by "true" self play executions, but I think there is so much training data discarded from MCTS simulation.

Quote from the paper:
```
All parameters of the model are trained jointly to accurately match the policy, value, and reward, for every
hypothetical step k, to corresponding target values observed after k actual time-steps have elapsed.
```

I checked this with a search for ".step" and there are only two positions in the source where it came up and they do not point to the tree search.

Please correct me if I get anything wrong about the paper or the source code.

Thanks

Sebastian

**Comment by vassago360** [comment]
I don't know if this helps you, but you may want to adjust number of simulations MCTS does.  It looks like with tic tac toe it does 25 simulations.  That's 25 node expansions.  I was surprised at how few that is.  I thought MCTS needs many simulations to get an average return to evaluate a node.

**Comment by Counterfeiter** [comment]
@vassago360 My main concern is not my slow training process it's the implementation of the training process here.

The main difference between AlphaZero and MuZero is that MuZero does not use a perfect game simulation while (self) playing, instead it uses the network outputs for the MCTS search.

But if MuZero trains it has access to a perfect game simulation. Training is evaluations of every MCTS search path with the given environment (see the quote from the paper above). In the current implementation the training data is only collected while self playing and only over the straight game until done, that slows down the training process.

A solution could be to deep copy the environment at each note to check different child nodes as well and collect obs, reward, actions...

Best regards

Sebastian

**Comment by fbondeaux** [comment]
@Counterfeiter Interesting thought. Have you implemented this yet?

**Comment by Counterfeiter** [comment]
@fbondeaux If I remember correctly I tried to store all MCTS searches ending in dones as training data. But I can't remember if it was this repo. But the results are worse... maybe to much bias for specific games. Maybe my quick test implementation fault. I don't know.

But I think I am wrong with the quote from above: The "hypotetical step k" is the fife step forward from the recurrent head with the learned model of the game and this is collected as training data, after the "real" env. is stepped through. 

I tried a lot with different alpha zero and muzero github repos. I finally changed the repo. "multiplayer-alphazero" for my needs. Now I can train any game. Not only zero sum, deterministic full information one or two player games... any game I can think of...

I am not really sure if muzero should take real game steps while traversing MCTS while training and use the model prediction only in the evaluation phase. That would make totally sense to my, but I didn't read the paper in depth and I haven't done any tests in this direction. Maybe you end up with a alpha zero implementation and alpha zeros trains a lot faster... So, as a result: it could be totally correct implemented, but isn't effective like alpha zero...

best regards

sebastian

**Comment by werner-duvaud** [comment]
Hi,
It seems to me that indeed you are discussing exactly the difference between Alpha zero and Muzero.
With Alpha zero the MCTS make calls to the environmentat each node but we must use deepcopy to revert the environement to a previous state.
In MuZero all the MCTS search happens "in MuZero mind/imagination" using only the learned model and without using exact information from the environment. This model is then improved once the action is taken in the environment.

A muzero that uses the real environment in the MCTS just during training seems a bit artificial but it should be doable :)
